File: backend/app/schemas/reader_chat.py
"""
Phase 10 reader-chat strict API and answer contracts.

Suggestions are chat data only (D-06): requires_explicit_confirmation is always
true and no confirmation/apply/write model exists in this phase.
"""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class StrictReaderChatModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class ConversationStatus(StrEnum):
    ACTIVE = "active"
    ARCHIVED = "archived"


class MessageRole(StrEnum):
    USER = "user"
    ASSISTANT = "assistant"


class GenerationJobStatus(StrEnum):
    QUEUED = "queued"
    RUNNING = "running"
    PAUSED_BUDGET = "paused_budget"
    PAUSED_DEPENDENCY = "paused_dependency"
    CANCELLED = "cancelled"
    COMPLETED = "completed"
    FAILED = "failed"
    FAILED_VALIDATION = "failed_validation"


class EvidenceSourceType(StrEnum):
    SELECTION = "selection"
    HIERARCHY = "hierarchy"
    TIMELINE = "timeline"
    KNOWLEDGE = "knowledge"
    RELATIONSHIP_OBSERVATION = "relationship_observation"


class SuggestionCandidateType(StrEnum):
    TIMELINE = "timeline"
    RELATIONSHIP = "relationship"
    CLUE = "clue"


# ---------------------------------------------------------------------------
# Selection and message request contracts
# ---------------------------------------------------------------------------


class SelectionCoordinate(StrictReaderChatModel):
    """Client-supplied selection; server re-slices Chapter.content as authority."""

    chapter_id: int = Field(gt=0)
    source_start: int = Field(ge=0)
    source_end: int = Field(gt=0)
    selection_text: str = Field(min_length=1, max_length=8000)
    selection_text_hash: str = Field(pattern=r"^[0-9a-f]{64}$")
    chapter_content_hash: str = Field(pattern=r"^[0-9a-f]{64}$")

    @model_validator(mode="after")
    def validate_half_open_range(self) -> SelectionCoordinate:
        if self.source_end <= self.source_start:
            raise ValueError("source_end must be greater than source_start")
        return self


class ChapterRange(StrictReaderChatModel):
    """Structure-anchored chapter interval (chapter-number semantics, inclusive).

    Matches timeline API chapter_start/chapter_end semantics. Server narrows
    chapter_end to the reading cutoff before any context assembly.
    """

    chapter_start: int = Field(ge=1)
    chapter_end: int = Field(ge=1)

    @model_validator(mode="after")
    def validate_inclusive_order(self) -> ChapterRange:
        if self.chapter_end < self.chapter_start:
            raise ValueError("chapter_end must be >= chapter_start")
        return self


class ChapterRangeAnchor(StrictReaderChatModel):
    """Echo of the effective (cutoff-narrowed) structure anchor on a message."""

    kind: Literal["chapter_range"] = "chapter_range"
    chapter_start: int = Field(ge=1)
    chapter_end: int = Field(ge=1)


class ConversationCreate(StrictReaderChatModel):
    title: str = Field(default="New chat", min_length=1, max_length=200)


class ConversationPatch(StrictReaderChatModel):
    title: str | None = Field(default=None, min_length=1, max_length=200)
    status: ConversationStatus | None = None

    @model_validator(mode="after")
    def require_at_least_one_field(self) -> ConversationPatch:
        if self.title is None and self.status is None:
            raise ValueError("at least one of title or status is required")
        return self


class ConversationListItem(StrictReaderChatModel):
    """Metadata-only list row — no message bodies or evidence excerpts."""

    id: int
    novel_id: int
    title: str
    status: ConversationStatus
    next_sequence: int
    last_opened_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    last_message_sequence: int | None = None
    last_message_role: MessageRole | None = None
    last_message_at: datetime | None = None


class ConversationDetail(ConversationListItem):
    pass


class MessageCreate(StrictReaderChatModel):
    client_message_id: str = Field(min_length=1, max_length=128)
    body: str = Field(min_length=1, max_length=8000)
    chapter_id: int | None = Field(default=None, gt=0)
    selection: SelectionCoordinate | None = None
    chapter_range: ChapterRange | None = None

    @model_validator(mode="after")
    def require_chapter_context(self) -> MessageCreate:
        if self.chapter_range is not None:
            # Structure anchor is exclusive with selection/single-chapter anchors.
            if self.selection is not None or self.chapter_id is not None:
                raise ValueError(
                    "chapter_range is mutually exclusive with selection and chapter_id"
                )
            return self
        if self.selection is None and self.chapter_id is None:
            raise ValueError(
                "chapter_id is required when selection and chapter_range are absent"
            )
        if (
            self.selection is not None
            and self.chapter_id is not None
            and self.selection.chapter_id != self.chapter_id
        ):
            raise ValueError("chapter_id must match selection.chapter_id")
        return self


class SelectionSummary(StrictReaderChatModel):
    chapter_id: int
    source_start: int
    source_end: int
    selection_text_hash: str
    chapter_content_hash: str


class CitationView(StrictReaderChatModel):
    block_id: str
    evidence_key: str
    context_evidence_ref_id: int
    chapter_id: int
    source_start: int
    source_end: int


class GenerationJobView(StrictReaderChatModel):
    id: int
    user_message_id: int
    status: GenerationJobStatus
    status_reason: str | None = None
    cancel_requested: bool
    retry_count: int
    error_code: str | None = None
    created_at: datetime
    updated_at: datetime


class MessageView(StrictReaderChatModel):
    id: int
    conversation_id: int
    sequence: int
    role: MessageRole
    body: str
    client_message_id: str | None = None
    reply_to_message_id: int | None = None
    selection: SelectionSummary | None = None
    anchor: ChapterRangeAnchor | None = None
    citations: list[CitationView] = Field(default_factory=list)
    generation_job: GenerationJobView | None = None
    queryplan: QueryPlanTraceView | None = None
    created_at: datetime


class QueryPlanTraceView(StrictReaderChatModel):
    """Trace/citation-level exposure shared by Reader and Analysis Chat (26-04).

    Mirrors the QueryPlan consumer view (trace id, availability, fallback and
    leaf citation-jump targets). Only leaf/raw evidence appears in
    ``citation_jump``; summaries, scores, routing metadata and chat text never
    do (D-08).
    """

    trace_id: str = Field(min_length=32, max_length=64)
    plan_hash: str = Field(pattern=r"^[0-9a-f]{64}$")
    intent: Literal["reader", "analysis"]
    anchor_kind: Literal["selection", "chapter_range"] | None = None
    cutoff_mode: str = Field(min_length=1, max_length=32)
    through_chapter: int = Field(gt=0)
    full_book_authorized: bool = False
    availability: list[dict[str, str]] = Field(default_factory=list)
    fallback: dict[str, Any] = Field(default_factory=dict)
    manifest_checksum: str = Field(pattern=r"^[0-9a-f]{64}$")
    allowed_evidence_ids: list[str] = Field(default_factory=list)
    citation_jump: list[dict[str, Any]] = Field(default_factory=list)
    abstained: bool = False


class MessageAccepted(StrictReaderChatModel):
    """202 response after durable user-message + job commit."""

    message: MessageView
    job: GenerationJobView


# ---------------------------------------------------------------------------
# Strict model answer envelope (AI-SPEC §5)
# ---------------------------------------------------------------------------


class AnswerBlock(StrictReaderChatModel):
    block_id: str = Field(min_length=1, max_length=80)
    text: str = Field(min_length=1, max_length=4000)
    evidence_refs: list[str] = Field(min_length=1, max_length=8)

    @field_validator("evidence_refs")
    @classmethod
    def non_empty_refs(cls, value: list[str]) -> list[str]:
        if not value:
            raise ValueError("answer blocks require at least one evidence ref")
        for ref in value:
            if not ref or not str(ref).strip():
                raise ValueError("evidence ref must be non-empty")
        return value


class UncertaintyPayload(StrictReaderChatModel):
    reason_code: str = Field(min_length=1, max_length=80)
    explanation: str = Field(min_length=1, max_length=2000)
    missing_evidence: list[str] = Field(default_factory=list)

    @model_validator(mode="before")
    @classmethod
    def normalize_known_reason_alias(cls, value: Any) -> Any:
        """Accept the provider's known ``reason`` alias without widening the schema."""

        if isinstance(value, dict) and "reason_code" not in value and "reason" in value:
            normalized = dict(value)
            normalized["reason_code"] = normalized.pop("reason")
            return normalized
        return value


class SuggestionCandidate(StrictReaderChatModel):
    """Chat-only proposal; never a domain write contract (D-06)."""

    candidate_type: SuggestionCandidateType
    target_ref: str | None = Field(default=None, max_length=160)
    proposal: str = Field(min_length=1, max_length=2000)
    evidence_refs: list[str] = Field(min_length=1, max_length=8)
    requires_explicit_confirmation: Literal[True] = True


class ReaderAnswerEnvelope(StrictReaderChatModel):
    schema_version: Literal["reader-answer.v1"] = "reader-answer.v1"
    answer_blocks: list[AnswerBlock] = Field(default_factory=list)
    clarifying_question: str | None = Field(default=None, min_length=1, max_length=1000)
    uncertainty: UncertaintyPayload | None = None
    suggestion_candidates: list[SuggestionCandidate] = Field(default_factory=list)

    @model_validator(mode="after")
    def require_content_or_uncertainty(self) -> ReaderAnswerEnvelope:
        if self.answer_blocks:
            return self
        if self.clarifying_question or self.uncertainty is not None:
            return self
        raise ValueError(
            "with empty answer_blocks, clarifying_question or uncertainty is required"
        )


def validate_answer_against_manifest(
    envelope: ReaderAnswerEnvelope,
    allowed_evidence_ids: set[str],
) -> None:
    """Business gate: every cited ref must belong to the frozen manifest allowlist."""

    def _check(refs: list[str], context: str) -> None:
        for ref in refs:
            if ref not in allowed_evidence_ids:
                raise ValueError(f"{context}: unknown evidence ref {ref!r}")

    for block in envelope.answer_blocks:
        _check(block.evidence_refs, f"answer block {block.block_id}")
    for suggestion in envelope.suggestion_candidates:
        _check(suggestion.evidence_refs, f"suggestion {suggestion.candidate_type}")


# ---------------------------------------------------------------------------
# Evidence / manifest wire shapes used by later plans
# ---------------------------------------------------------------------------


class ContextEvidenceRefView(StrictReaderChatModel):
    evidence_key: str
    source_type: EvidenceSourceType
    source_id: str
    chapter_id: int
    chapter_number: int
    source_start: int
    source_end: int
    content_hash: str
    excerpt: str
    sort_order: int = 0
    version_lineage: dict[str, Any] = Field(default_factory=dict)


class ContextManifestView(StrictReaderChatModel):
    id: int
    user_message_id: int
    full_book: bool
    cutoff_chapter_number: int
    manifest_checksum: str
    reading_progress_snapshot: dict[str, Any]
    evidence_refs: list[ContextEvidenceRefView] = Field(default_factory=list)

File: frontend/src/lib/api/reader-chat.ts
/**
 * 阅读器选区对话 API（Phase 10）。
 */

import { api } from "./client";

export type ConversationStatus = "active" | "archived";
export type MessageRole = "user" | "assistant";

export type GenerationJobStatus =
  | "queued"
  | "running"
  | "paused_budget"
  | "paused_dependency"
  | "cancelled"
  | "completed"
  | "failed"
  | "failed_validation";

export interface SelectionCoordinate {
  chapter_id: number;
  source_start: number;
  source_end: number;
  selection_text: string;
  selection_text_hash: string;
  chapter_content_hash: string;
}

export interface SelectionSummary {
  chapter_id: number;
  source_start: number;
  source_end: number;
  selection_text_hash: string;
  chapter_content_hash: string;
}

export interface CitationView {
  block_id: string;
  evidence_key: string;
  context_evidence_ref_id: number;
  chapter_id: number;
  source_start: number;
  source_end: number;
}

export interface GenerationJobView {
  id: number;
  user_message_id: number;
  status: GenerationJobStatus;
  status_reason: string | null;
  cancel_requested: boolean;
  retry_count: number;
  error_code: string | null;
  created_at: string;
  updated_at: string;
}

export interface ConversationListItem {
  id: number;
  novel_id: number;
  title: string;
  status: ConversationStatus;
  next_sequence: number;
  last_opened_at: string | null;
  created_at: string;
  updated_at: string;
  last_message_sequence: number | null;
  last_message_role: MessageRole | null;
  last_message_at: string | null;
}

export type ConversationDetail = ConversationListItem;

/** 25.1-01：结构区间锚点回显（服务端已按剧透边界收窄后的实际区间）。 */
export interface ChapterRangeAnchor {
  kind: "chapter_range";
  chapter_start: number;
  chapter_end: number;
}

/** 26-04：共享 QueryPlan trace/citation 暴露（Reader/Analysis Chat 同一核心）。 */
export interface QueryPlanTraceView {
  trace_id: string;
  plan_hash: string;
  intent: "reader" | "analysis";
  anchor_kind: "selection" | "chapter_range" | null;
  cutoff_mode: string;
  through_chapter: number;
  full_book_authorized: boolean;
  availability: Array<Record<string, string>>;
  fallback: Record<string, unknown>;
  manifest_checksum: string;
  allowed_evidence_ids: string[];
  citation_jump: Array<{
    evidence_key: string;
    chapter_id: number;
    chapter_number: number;
    source_start: number;
    source_end: number;
    excerpt: string;
  }>;
  abstained: boolean;
  /** 27-04：world projection authority/disclosure/evidence 契约（可缺省）。 */
  world_projection?: WorldProjectionView | null;
}

/**
 * 27-04：World projection 序列化契约（后端 queryplan/contracts.py 镜像）。
 * authority 严格四标签之一；candidate-only 经 approved 标识；user
 * interpretation 隔离在 overrides 中（D-06）。
 */
export type WorldAuthorityLabel =
  | "canon_fact"
  | "probable_inference"
  | "literary_interpretation"
  | "user_interpretation";

export interface WorldProjectionItemView {
  claim_key: string;
  kind: "character" | "world";
  subject: string;
  aspect: string;
  proposition: string;
  authority: WorldAuthorityLabel;
  known_at: number;
  disclosure_cutoff: number;
  pov: string | null;
  gate_status: "pending" | "passed" | "rejected";
  approved: boolean;
  is_override: boolean;
  evidence_key: string;
  chapter_id: number;
  chapter_number: number;
  source_start: number;
  source_end: number;
  content_hash: string;
  source_snapshot_hash: string;
  lineage: string[];
}

export interface WorldProjectionView {
  schema_version: string;
  available: boolean;
  status: "available" | "candidate_only" | "unavailable";
  cutoff: number;
  items: WorldProjectionItemView[];
  overrides: WorldProjectionItemView[];
  authorities: WorldAuthorityLabel[];
  manifest_checksum?: string | null;
  snapshot_hash?: string | null;
}

export interface MessageView {
  id: number;
  conversation_id: number;
  sequence: number;
  role: MessageRole;
  body: string;
  client_message_id: string | null;
  reply_to_message_id: number | null;
  selection: SelectionSummary | null;
  anchor?: ChapterRangeAnchor | null;
  citations: CitationView[];
  generation_job: GenerationJobView | null;
  queryplan?: QueryPlanTraceView | null;
  backfill_runs?: BackfillRunView[];
  created_at: string;
}

export interface BackfillRunView {
  run_id: number;
  skill_name: string;
  status: "queued" | "running" | "cancelled" | "completed" | "failed";
  backfill_dimension?: string | null;
}

export interface MessageAccepted {
  message: MessageView;
  job: GenerationJobView;
}

export interface ConversationListResponse {
  items: ConversationListItem[];
  total: number;
  skip: number;
  limit: number;
}

export interface MessageListResponse {
  items: MessageView[];
  total: number;
  skip: number;
  limit: number;
  after_sequence: number;
}

export interface MessageCreateBody {
  client_message_id: string;
  body: string;
  chapter_id?: number;
  selection?: SelectionCoordinate;
  /** 与 chapter_id/selection 互斥；章号语义，服务端按剧透边界收窄 chapter_end。 */
  chapter_range?: { chapter_start: number; chapter_end: number };
}

const TERMINAL_JOB_STATUSES: ReadonlySet<GenerationJobStatus> = new Set([
  "completed",
  "cancelled",
  "failed",
  "failed_validation",
  "paused_budget",
  "paused_dependency",
]);

export function isTerminalJobStatus(status: string): boolean {
  return TERMINAL_JOB_STATUSES.has(status as GenerationJobStatus);
}

/**
 * Poll a generation job until terminal or timeout.
 * Never fabricates assistant content — only returns server job views.
 */
export async function pollReaderChatJob(
  novelId: string | number,
  conversationId: number,
  jobId: number,
  options?: {
    intervalMs?: number;
    timeoutMs?: number;
    signal?: AbortSignal;
    onUpdate?: (job: GenerationJobView) => void;
  }
): Promise<GenerationJobView> {
  const intervalMs = options?.intervalMs ?? 800;
  const timeoutMs = options?.timeoutMs ?? 120_000;
  const started = Date.now();
  let last: GenerationJobView | null = null;
  while (Date.now() - started < timeoutMs) {
    if (options?.signal?.aborted) {
      throw new DOMException("Aborted", "AbortError");
    }
    const res = await readerChatApi.getJob(novelId, conversationId, jobId);
    last = res.data;
    options?.onUpdate?.(last);
    if (isTerminalJobStatus(last.status)) return last;
    await new Promise((r) => setTimeout(r, intervalMs));
  }
  if (last) return last;
  throw new Error("job poll timeout with no response");
}

export const readerChatApi = {
  listConversations: (
    novelId: string | number,
    params?: { status?: ConversationStatus; skip?: number; limit?: number }
  ) =>
    api.get<ConversationListResponse>(`/novels/${novelId}/conversations`, {
      params: {
        status: params?.status,
        skip: params?.skip,
        limit: params?.limit,
      },
    }),

  createConversation: (novelId: string | number, title = "New chat") =>
    api.post<ConversationDetail>(`/novels/${novelId}/conversations`, { title }),

  getConversation: (novelId: string | number, conversationId: number) =>
    api.get<ConversationDetail>(
      `/novels/${novelId}/conversations/${conversationId}`
    ),

  patchConversation: (
    novelId: string | number,
    conversationId: number,
    body: { title?: string; status?: ConversationStatus }
  ) =>
    api.patch<ConversationDetail>(
      `/novels/${novelId}/conversations/${conversationId}`,
      body
    ),

  deleteConversation: (novelId: string | number, conversationId: number) =>
    api.delete(`/novels/${novelId}/conversations/${conversationId}`),

  listMessages: (
    novelId: string | number,
    conversationId: number,
    params?: { after_sequence?: number; skip?: number; limit?: number }
  ) =>
    api.get<MessageListResponse>(
      `/novels/${novelId}/conversations/${conversationId}/messages`,
      {
        params: {
          after_sequence: params?.after_sequence,
          skip: params?.skip,
          limit: params?.limit,
        },
      }
    ),

  createMessage: (
    novelId: string | number,
    conversationId: number,
    body: MessageCreateBody
  ) =>
    api.post<MessageAccepted>(
      `/novels/${novelId}/conversations/${conversationId}/messages`,
      body
    ),

  getJob: (novelId: string | number, conversationId: number, jobId: number) =>
    api.get<GenerationJobView>(
      `/novels/${novelId}/conversations/${conversationId}/jobs/${jobId}`
    ),

  cancelJob: (
    novelId: string | number,
    conversationId: number,
    jobId: number
  ) =>
    api.post<GenerationJobView>(
      `/novels/${novelId}/conversations/${conversationId}/jobs/${jobId}/cancel`
    ),

  retryJob: (
    novelId: string | number,
    conversationId: number,
    jobId: number
  ) =>
    api.post<GenerationJobView>(
      `/novels/${novelId}/conversations/${conversationId}/jobs/${jobId}/retry`
    ),
};
